lapce-data/src/editor/commands/insert_chars.rs

```diff
@@ -0,0 +1,291 @@
+                if buffer.buffer.slice_to_cow(line_start..offset).trim() == "" {
+                    let opening_character = matching_char(c).unwrap();
+                    if let Some(previous_offset) = buffer.buffer.previous_unmatched(
+                        syntax.as_ref(),
+                        opening_character,
+                        offset,
+                    ) {
+                        // Auto-indent closing character to the same level as the opening.
+                        let previous_line =
+                            buffer.buffer.line_of_offset(previous_offset);
+                        let line_indent =
+                            buffer.buffer.indent_on_line(previous_line);
+
+                        let current_selection = region_to_selection(SelRegion::new(
+                            line_start, offset, None,
+                        ));
+
+                        edits.push((current_selection, format!("{line_indent}{c}")));
+                        continue;
+                    }
+                }
+            }
+
+            if matching_pair_type == Some(true) {
+                // Create a late edit to insert the closing pair, if allowed.
+                let is_whitespace_or_punct = cursor_char
+                    .map(|c| {
+                        let prop = get_word_property(c);
+                        prop == WordProperty::Lf
+                            || prop == WordProperty::Space
+                            || prop == WordProperty::Punctuation
+                    })
+                    .unwrap_or(true);
+
+                if is_whitespace_or_punct {
+                    let insert_after = matching_char(c).unwrap();
+                    edits_after.push((idx, insert_after));
+                }
+            };
+
+            let current_selection = region_to_selection(*region);
+
+            edits.push((current_selection, c.to_string()));
+        }
+
+        // Apply edits to current selection
+        let edits = edits
+            .iter()
+            .map(|(selection, content)| (selection, content.as_str()))
+            .collect::<Vec<_>>();
+
+        let delta = buffer.edit_multiple(&edits, EditType::InsertChars);
+
+        // Update selection
+        let mut selection =
+            selection.apply_delta(&delta, true, InsertDrift::Default);
+
+        // Apply late edits
+        let edits_after = edits_after
+            .iter()
+            .map(|(idx, content)| {
+                (
+                    region_to_selection(selection.regions()[*idx]),
+                    content.to_string(),
+                )
+            })
+            .collect::<Vec<_>>();
+
+        let edits_after = edits_after
+            .iter()
+            .map(|(selection, content)| (selection, content.as_str()))
+            .collect::<Vec<_>>();
+
+        buffer.edit_multiple(&edits_after, EditType::InsertChars);
+
+        // Adjust selection according to previous late edits
```

Comment:
@dzhou121 while this function got more complex, this (L155-175) is the part that was missing previously.